### PR TITLE
EOS-17974: use pool version Id in read/write of objects

### PR DESCRIPTION
**Pull request by evgeniy-brazhnikov**
Signed-off-by: evgeniy-brazhnikov <evgeniy.brazhnikov@seagate.com>

-----
[View rendered docs/design/pool-version-id.md](https://github.com/Seagate/cortx-s3server/blob/br/eb/eos-17974/docs/design/pool-version-id.md)

### CODE REVIEW

File: docs/design/pool-version-id.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+Contents
+
+- [Pool version Id](#pool-version-id)
+  - [License](#license)
+  - [Proposed Design Change](#proposed-design-change)
+
+
+## License
+
+Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
```
Comment (knrajnambiar76) [inline docs/design/pool-version-id.md:12]: 2021

File: docs/design/pool-version-id.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+## Proposed Design Change
+
+This document briefly describes proposed change - pool version id usage.
+
+Pool version is an object in configuration DB of Motr, which describes a subset of cluster.
+It contains disks, controllers of disks and racks for controllers.
+Cluster can be displayed as a tree with physical disks as leafs.
+A pool version is a subtree of that tree.
+Any Motr objects and indices are created and stored in a single pool version which is selected when an entity is creating.
```
Comment (knrajnambiar76) [inline docs/design/pool-version-id.md:38]: created.

File: docs/design/pool-version-id.md (hunk 1 above)
Comment (evgeniy-brazhnikov) [reply docs/design/pool-version-id.md]: corrected

File: docs/design/pool-version-id.md (hunk 2 above)
Comment (evgeniy-brazhnikov) [reply docs/design/pool-version-id.md]: corrected

File: server/s3_put_object_action_base.cc [hunk 3]
```diff
@@ -467,11 +465,12 @@ void S3PutObjectActionBase::delete_old_object() {
   // If PUT is success, we delete old object if present
   assert(old_object_oid.u_hi != 0ULL || old_object_oid.u_lo != 0ULL);
 
-  motr_writer->set_oid(old_object_oid);
   motr_writer->delete_object(
       std::bind(&S3PutObjectActionBase::remove_old_object_version_metadata,
                 this),
-      std::bind(&S3PutObjectActionBase::next, this), old_layout_id);
+      std::bind(&S3PutObjectActionBase::next, this), old_object_oid,
```
Comment (t7ko-seagate) [inline server/s3_put_object_action_base.cc:471]: this is happy path, so need to specify PVID here as well.

File: ut/s3_put_chunk_upload_object_action_test.cc [hunk 4]
```diff
@@ -953,7 +950,6 @@ TEST_F(S3PutChunkUploadObjectActionTestNoAuth, SaveMetadata) {
   EXPECT_CALL(*(motr_writer_factory->mock_motr_writer), get_oid())
       .Times(AtLeast(1))
       .WillOnce(Return(oid));
-  EXPECT_CALL(*(object_meta_factory->mock_object_metadata), set_oid(_))
```
Comment (t7ko-seagate) [inline ut/s3_put_chunk_upload_object_action_test.cc:956]: this looks like an error, this should not even compile, as removing this line makes the next one syntax-wrong:

```
.Times(AtLeast(1));
```

**Review by t7ko-seagate — APPROVED** [review]
Change is ok, but not 100% complete, as there are few happy-path scenarios not yet addressed.  So can merge now and handle remaining scenarios in next sprint, or wait this PR till all scenarios addressed.

**Review by knrajnambiar76 — APPROVED** [review]

File: server/s3_motr_writer.h [hunk 5]
```diff
@@ -190,22 +190,24 @@ class S3MotrWiter {
...
+                             std::function<void(void)> on_failed,
+                             const struct m0_uint128& object_id, int layoutid);
   // Async save operation.
   virtual void write_content(std::function<void(void)> on_success,
                              std::function<void(void)> on_failed,
                              S3BufferSequence buffer_sequence,
                              size_t size_of_each_buf);
 
   // Async delete operation.
+  // TODO: add pool version id into BackgroundDelete memo
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_motr_writer.h:202]: No. As i understand after quick look, BG delete (Python) code shall be changed too.

File: ut/s3_put_chunk_upload_object_action_test.cc (hunk 4 above)
Comment (evgeniy-brazhnikov) [reply ut/s3_put_chunk_upload_object_action_test.cc:956]: It's inside commented snippet. Corrected

**Comment by evgeniy-brazhnikov** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/1070/

**Comment by evgeniy-brazhnikov** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/1078/

File: docs/design/pool-version-id.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# Pool version Id
+
+Contents
+
+- [Pool version Id](#pool-version-id)
+  - [License](#license)
+  - [Proposed Design Change](#proposed-design-change)
+
+
+## License
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply docs/design/pool-version-id.md]: Done

File: server/s3_m0_uint128_helper.cc [hunk 7]
```diff
@@ -62,4 +63,22 @@ m0_uint128 to_m0_uint128(const std::string &id_str) {
   return id;
 }
 
+void to_string(const struct m0_fid &fid, std::string &s_res) {
+  s_res = base64_encode(reinterpret_cast<unsigned char const *>(&fid),
+                        sizeof(struct m0_fid));
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_m0_uint128_helper.cc:70]: Done

File: server/s3_object_metadata.cc [hunk 8]
```diff
@@ -778,3 +779,16 @@ bool S3ObjectMetadata::check_object_tags_exists() {
...
+  struct m0_fid pvid;
+  S3M0Uint128Helper::to_m0_fid(pvid_str, pvid);
+  return pvid;
+}
+
+void S3ObjectMetadata::set_pvid(const struct m0_fid* p_pvid) {
+  if (p_pvid) {
+    S3M0Uint128Helper::to_string(*p_pvid, pvid_str);
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id, "%s - NULL pointer", __func__);
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_object_metadata.cc:792]: Done

File: server/s3_post_complete_action.cc [hunk 9]
```diff
@@ -918,11 +918,12 @@ void S3PostCompleteAction::delete_old_object() {
     next();
     return;
   }
-  motr_writer->set_oid(old_object_oid);
   motr_writer->delete_object(
       std::bind(&S3PostCompleteAction::remove_old_object_version_metadata,
                 this),
-      std::bind(&S3PostCompleteAction::next, this), old_layout_id);
+      std::bind(&S3PostCompleteAction::next, this), old_object_oid,
+      old_layout_id);
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_post_complete_action.cc]: Done

File: server/s3_put_chunk_upload_object_action.cc [hunk 10]
```diff
@@ -990,15 +989,14 @@ void S3PutChunkUploadObjectAction::delete_old_object() {
            "BD.\n");
     // Call next task in the pipeline
     next();
-    return;
+  } else {
+    motr_writer->delete_object(
+        std::bind(
+            &S3PutChunkUploadObjectAction::remove_old_object_version_metadata,
+            this),
+        std::bind(&S3PutChunkUploadObjectAction::next, this), old_object_oid,
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_put_chunk_upload_object_action.cc:997]: Done

File: server/s3_put_object_action_base.cc (hunk 3 above)
Comment (evgeniy-brazhnikov) [reply server/s3_put_object_action_base.cc:471]: Done

**Pull request merged**